### PR TITLE
mlx: rework the MLX sampler

### PR DESCRIPTION
**Pull request by pdevine**
Replace the MLX sampler transform chain with an explicit distribution pipeline that applies:
  1. penalties
  2. top-k
  3. temperature/softmax
  4. top-p
  5. min-p
  6. normalize
  7. categorical

The common top_k path now keeps sparse [B,K] token ids/probabilities on GPU instead of carrying full-vocab scores, and sampled MTP reuses those draft/target distributions for acceptance, bonus, and residual sampling.

This change also fixes the seed parameter so that temperature sampling and sampled MTP are reproducible.

### CODE REVIEW

File: x/mlxrunner/sample/sample.go [hunk 1]
```diff
@@ -627,74 +764,107 @@ func (s *Sampler) sampleTokensSerial(slots []*slotState, logits *mlx.Array) *mlx
...
+		return Distribution{IDs: ids, Probs: probs}
 	}
 
-	maxScore := scores.MaxAxis(-1, true)
-	threshold := mlx.AddScalar(maxScore, float32(math.Log(float64(ctx.opts.MinP))))
+	vocab := scores.Dim(scores.NumDims() - 1)
+	if slot.opts.TopK > 0 && slot.opts.TopK < vocab {
+		return sparseDistribution(ctx.opts, scores)
+	}
+	return denseDistribution(ctx.opts, scores)
```
Comment (ParthSareen) [inline x/mlxrunner/sample/sample.go:820]: Is there any case where we need the entire dense branch? We should always have some topK. Do we currently let people turn it off in the existing samplers?

File: x/mlxrunner/sample/sample.go [hunk 2]
```diff
@@ -513,6 +649,9 @@ func (s *Sampler) canBatch(slots []*slotState) (Options, bool) {
 	// slots is non-empty (Sample guards) and every slot is registered,
 	// so s.slots[0].opts is the canonical shared value.
 	shared := s.slots[0].opts
+	// TODO: Before using multi-slot batching with seeded stochastic sampling,
```
Comment (ParthSareen) [inline x/mlxrunner/sample/sample.go]:
```suggestion
	// TODO(pdevine): Before using multi-slot batching with seeded stochastic sampling,
```

File: x/mlxrunner/sample/sample.go [hunk 3]
```diff
@@ -627,74 +764,107 @@ func (s *Sampler) sampleTokensSerial(slots []*slotState, logits *mlx.Array) *mlx
...
+func (slot *slotState) sample(ctx *slotCtx, logits *mlx.Array) *mlx.Array {
+	if slot.opts.Temperature == 0 {
+		return slot.baseScores(ctx, logits).Argmax(-1, false).AsType(mlx.DTypeInt32)
+	}
+	return slot.distribution(ctx, logits).SampleWithKey(slot.nextRandomKey())
 }
 
-func temperature(ctx *slotCtx, scores *mlx.Array) *mlx.Array {
-	return mlx.DivScalar(scores, ctx.opts.Temperature).Categorical(-1).AsType(mlx.DTypeInt32)
+func (slot *slotState) nextRandomKey() *mlx.Array {
```
Comment (ParthSareen) [inline x/mlxrunner/sample/sample.go:774]: Is this all to help with stochastic sampling? Does this need to be so manual?

File: x/mlxrunner/sample/sample.go (hunk 1 above)
Comment (pdevine) [reply x/mlxrunner/sample/sample.go:820]: We currently support full vocab sampling in both the ggml + llama-server runners. It's also supported in sglang and vllm. You can turn that on with topK == 0|-1

File: x/mlxrunner/sample/sample.go (hunk 3 above)
Comment (pdevine) [reply x/mlxrunner/sample/sample.go:774]: Yes. And specifically to let the RNG work w/ the seed properly (which didn't work up until this point). This specific part is to leave the breadcrumb for batch sampling.

**Review by ParthSareen — APPROVED** [review]
